### ISSUE TITLE
Alias definition before class weirdness with nilable type not functioning as optional field

### ISSUE THREAD

**Issue by PennyJim** [EmmyLuaLs/emmylua-analyzer-rust#1135]
I am taking a Factorio mod that was typed for LuaLS, and moving its types over to emmylua.
The first mod I am working on has a structure that may or may not contain a type that is itself recursive (with `nil` being a possible value).

I found that my tables were complaining about it missing the "required" field, and when investigating a simple case I found that whether or not the type has already been defined by the point of the class definition seems to determine the nilability behavior.

Instance 1:
```lua
---@alias B true?
---@class A
---@field field? B

---@type A
local var = {} -- Warning: Missing Required fields in type `A`: `field

---@type A
local var = {field = {}} -- No warning (wrong)

---@type A
local var = {field = "bleh"} -- Warning: Cannot assign `string` to `B`.
```

Instance 2:
```lua
---@class A
---@field field? B
---@alias B true?

---@type A
local var = {} -- No warning (correct)

---@type A
local var = {field = {}} -- Warning: Cannot assign `table` to `B?`

---@type A
local var = {field = "bleh"} -- Warning: Cannot assign `string` to `B?`
```

And more visible deeper insight (probably)
Instance 3:
```lua
---@alias B number|true
---@class A
---@field field? B

---@type A
local var = {field = {}} -- Warning: Cannot assign `table` to `(number|true)?`.
```

Instance 4:
```lua
---@class A
---@field field? B
---@alias B number|true

---@type A
local var = {field = {}} -- Warning: Cannot assign `table` to `B?`.
```

**Comment by PennyJim** [comment]
You made me worried because the test you put in the latest commit only covered the `{field = {}}` case.

Can confirm it fixes the warning on a "required" optional field :)
